_posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc

```diff
@@ -0,0 +1,2334 @@
+  talk
+  https://blog.gruntwork.io/new-talk-automated-testing-for-terraform-docker-packer-kubernetes-and-more-cba312171aa6[Automated
+  testing for Terraform, Docker, Packer, Kubernetes, and More] and blog post
+  https://www.ybrikman.com/writing/2016/02/14/agility-requires-safety/[Agility requires safety].
+
+
+[[core_concepts]]
+== Core Concepts
+
+[[why_is_it_important_to_have_cicd]]
+=== Why is it important to have CI/CD?
+
+To understand the benefits of CI/CD, it is worth exploring the opposite: _late integration and late delivery (LI/LD)_.
+We will explain LI/LD using a thought experiment about building the International Space Station (ISS).
+
+.Components of the International Space Station
+image::/assets/img/guides/infrastructure-cicd-pipeline/iss-components.png[]
+
+The ISS consists of dozens of components, as shown in the image above. Each component is built by a team from a
+different country, with a central team responsible for organizing development. In LI/LD,
+you would organize the development by designing all the components in advance, and then have each team go
+off and work on their component in _total isolation_. There is complete trust in the design and the teams such that
+there is no need to check in and integrate the components along the way. When all the teams are done, each team launches
+the component into space and then put all the components together at the same time in space, _for the first time_.
+
+It isn't hard to imagine that this could be disastrous: one team would think the other team was responsible for wiring
+while that team thought everything would be wireless; all the teams would use the metric system except one; everyone cut
+toilets from the scope thinking some other team is sure to include it. Finding all of this out once everything has
+already been built and is floating in outer space means that fixing the problems will be very difficult and expensive.
+
+While it is hard to imagine that anyone would build the ISS in this way, unfortunately this model of development is
+fairly common in the software industry. Developers work in total isolation for weeks or months at a time on _feature
+branches_ without integrating their work with other teams, and then try to merge all the work together at the last
+minute moments before release. As a result, the integration process is very expensive and takes days or weeks fixing merge
+conflicts, tracking down subtle bugs, and trying to stabilize release branches.
+
+.Many teams employ a practice of working on their features over long periods of time on isolated branches. These long lived feature branches have higher chances of having big conflicts to resolve when integrated.
+image::/assets/img/guides/infrastructure-cicd-pipeline/feature-branch-merge-conflict.png[]
+
+In contrast, the Continuous Integration and Continuous Delivery model of development promotes more cross team
```

Comment:
In the asciidoc rendered view (on the website) it will look ok. See how the caption is rendered here:

<img width="787" alt="How_to_deploy_a_production-grade_Kubernetes_cluster_on_AWS_-_Mozilla_Firefox" src="https://user-images.githubusercontent.com/430092/76339997-17bd2e00-62b8-11ea-97b2-1d8aa4aec5ea.png">